builder/builder.go

```diff
@@ -1916,3 +1928,99 @@ func writeMetaFiles(path, format, version string) error {
 
 	return ioutil.WriteFile(filepath.Join(path, "mixer-src-version"), []byte(version), 0644)
 }
+
+func (b *Builder) getLatestUpstreamSameFormat() (format string, latest uint32, err error) {
+	format, err = b.DownloadFileFromUpstream(fmt.Sprintf("update/%d/format", b.UpstreamVersion))
+	if err != nil {
+		return "", 0, errors.Wrap(err, "couldn't download information about upstream")
+	}
+
+	latestStr, err := b.DownloadFileFromUpstream(fmt.Sprintf("update/version/format%s/latest", format))
+	if err != nil {
+		return "", 0, errors.Wrap(err, "couldn't download information about upstream")
+	}
+
+	latest, err = parseUint32(latestStr)
+	if err != nil {
+		return "", 0, errors.Wrap(err, "couldn't parse latest upstream version")
+	}
+
+	return format, latest, err
+}
+
+// PrintVersions prints the current mix and upstream versions, and the
+// latest version of upstream.
+func (b *Builder) PrintVersions() error {
+	format, latest, err := b.getLatestUpstreamSameFormat()
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf(`Current mix:      %d
+Current upstream: %d (format: %s)
+
+Latest upstream:  %d (format: %s)
+`, b.MixVersion, b.UpstreamVersion, format, latest, format)
+
+	return nil
+}
+
+// UpdateVersions will validate then update both mix and upstream versions. If upstream
+// version is 0, then the latest upstream version possible will be taken instead.
+func (b *Builder) UpdateVersions(nextMix, nextUpstream uint32) error {
+	format, latest, err := b.getLatestUpstreamSameFormat()
+	if err != nil {
+		return err
+	}
+
+	if nextMix <= b.MixVersion {
+		return fmt.Errorf("invalid mix version to update (%d), need to be greater than current mix version (%d)", nextMix, b.MixVersion)
+	}
+
+	if nextUpstream < b.UpstreamVersion {
+		return fmt.Errorf("invalid upstream version to update (%d), need to be at least the same as current upstream version (%d)", nextUpstream, b.UpstreamVersion)
+	}
+
+	switch {
+	case nextUpstream == 0:
+		nextUpstream = latest
+
+	case nextUpstream < b.UpstreamVersion:
```

Comment:
Perhaps I'm missing something, but is this not the same check as L1980 above?

Reply:
What if I started with if then switched to switch?

Fixed.